Add toIP to CiliumLocalRedirectPolicy redirectBackend (#41645)
cilium-team/cilium: Documentation/network/kubernetes/local-redirect-policy.rst

            $ kubectl exec -it cilium-mhnhz -n kube-system -- cilium-dbg service list | grep LocalRedirect
            11   10.96.0.10:53      LocalRedirect   1 => 10.244.1.49:53 (active)

GKE Metadata Server with OverrideIP

> ------------------------------------

Reviewer: ```suggestion
GKE Metadata Server with OverrideIP
-----------------------------------
```

The GKE metadata server is a pod running in the host network namespace. The new version of GKE metadata server only

> listens a specific secondary address from the loopback interface. GKE uses the IP table rule to redirect the traffic

Reviewer: ```suggestion
listens on a specific secondary address from the loopback interface. GKE uses the IP table rule to redirect the traffic
```

to the metadata server.

If you are using the Cilium eBPF host routing, the traffic will skip the IP table rule and leave the node through the host interface. 
To optimize the traffic path, you can use the local redirect policy with ``overrideIP`` field to redirect the traffic to the metadata server 
pod on the local node with Cilium eBPF host routing.

> Assume you have the following helm values for Cilium:

Reviewer: ```suggestion
Assuming you have the following Helm values for Cilium:
```

.. code-block:: shell-session

   bpf:
     masquerade: true
   kubeProxyReplacement: true
   localRedirectPolicies:

>      enabled: true

Reviewer: ```suggestion
.. code-block:: yaml

   bpf:
     masquerade: true
   kubeProxyReplacement: true
   localRedirectPolicies:
     enabled: true
```

> The GKE metadata server pod is deployed and bind to the ``169.254.169.252:988``.

Reviewer: ```suggestion
The GKE metadata server pod is deployed and binds to ``169.254.169.252:988``.
```

    03:56:57.652535 IP 169.254.169.254.80 > 172.16.64.8.47478: Flags [S.], seq 1674047236, ack 2476202556, win 65535, options [mss 1420,eol], length 0
    03:56:57.652643 IP 172.16.64.8.47478 > 169.254.169.254.80: Flags [.], ack 1, win 42600, length 0

You can use the following local redirect policy to redirect the traffic to the metadata server pod on the local node:

.. code-block:: yaml

    apiVersion: cilium.io/v2
    kind: CiliumLocalRedirectPolicy
    metadata:
      name: gke-metadata-server
      namespace: kube-system
    spec:
      redirectBackend:
        overrideIP: "169.254.169.252"
        localEndpointSelector:
          matchLabels:
            k8s-app: "gke-metadata-server"
        toPorts:
        - port: "988"
          protocol: "TCP"
      redirectFrontend:
        addressMatcher:
          ip: "169.254.169.254"
          toPorts:
          - port: "80"
            protocol: "TCP"

After applying the local redirect policy, the traffic to the metadata server will be redirected to the node-local backend pod.

> You can check local redirect policy to see the policy has been applied from cilium agent through the following commands:

Reviewer: ```suggestion
You can check local redirect policy to see the policy has been applied from the Cilium agent with the following command:
```

.. code-block:: shell-session

    $ cilium-dbg service list | grep LocalRedirect
    19   169.254.169.254:80/TCP   LocalRedirect   1 => 169.254.169.252:988/TCP (active)

> 
> 
> 

Reviewer: ```suggestion

```

(one blank line is enough, you've got three)

You can use ``tcpdump`` on the host to verify the traffic is redirected to the node-local backend pod.

> As you can see from the following output the packet is sent to the node-local backend pod through ``lxc2140a2364964`` directly from the pod(10.36.1.28).

Reviewer: ```suggestion
As you can see from the following output the packet is sent to the node-local backend pod through ``lxc2140a2364964`` directly from the pod (``10.36.1.28``).
```

    In order to enable this configuration starting Cilium version 1.16.0, previously applied local redirect policies
    and policies selected backend pods need to be deleted, and re-created.

Override IP
-----------
The local redirect policy supports overriding the IP address of the backend pod.
This can be used to redirect traffic to a different IP address than the one assigned by Kubernetes.

> This is useful when the backend pod is running in the host network namespace where it could only listen on the specific address from loopback interface(such as GKE metadata server).

Reviewer: Nit: missing a space before the parenthesis

```suggestion
This is useful when the backend pod is running in the host network namespace where it could only
listen on the specific address from loopback interface (such as GKE metadata server).
```

> you can use the ``overrideIP`` field to override the IP address of the backend pod.

Reviewer: ```suggestion
You can use the ``overrideIP`` field to override the IP address of the backend pod.
```

.. code-block:: yaml

    apiVersion: cilium.io/v2
    kind: CiliumLocalRedirectPolicy
    metadata:
      name: lrp-addr
      namespace: default
    spec:
      redirectBackend:
        localEndpointSelector:
          matchLabels:
            app: proxy
        toPorts:
          - port: "50"
            name: "test"
            protocol: TCP
            overrideIP: 1.1.1.3
          - port: "51"
            name: "test1"
            protocol: TCP
            overrideIP: 1.1.1.4

.. note::

>     Even though you can configure the ``overrideIP`` field with ``redirectFrontend`` in the policy, it will not be used.

Reviewer: I find the phrasing could be improved. What will not be used in that case, `overrideIP` or `redirectFrontend`? Maybe simply:

```suggestion
    Field ``overrideIP`` has no effect when used for ``redirectFrontend``.
```

Author: I think yours is accurate.

    In order to enable this configuration starting Cilium version 1.16.0, previously applied local redirect policies
    and policies selected backend pods need to be deleted, and re-created.

Override IP
-----------
The local redirect policy supports overriding the IP address of the backend pod.
This can be used to redirect traffic to a different IP address than the one assigned by Kubernetes.
This is useful when the backend pod is running in the host network namespace where it could only 

> listen on the specific address from loopback interface (such as GKE metadata server).

Reviewer 2: Can you add an example with GKE use case under this section - https://docs.cilium.io/en/latest/network/kubernetes/local-redirect-policy/#use-cases?

Author: sure. I will add that later.
